src/Gui/Hooks/useStoryPreview/StoryPreviewer.luau

```diff
@@ -0,0 +1,181 @@
+--!nonstrict
+local ERROR_DELAY = 1
+
+local StoryPreviewer = {}
+StoryPreviewer.__index = StoryPreviewer
+
+function StoryPreviewer.new(rootRef, setCurrentPreview)
+	local self = setmetatable({}, StoryPreviewer)
+
+	self.errorId = 0
+
+	local display = Instance.new("ScreenGui")
+	display.Name = "HoarcekatDisplay"
+	display.ZIndexBehavior = Enum.ZIndexBehavior.Sibling
+	self.display = display
+
+	self.rootRef = rootRef
+	self.setCurrentPreview = setCurrentPreview
+
+	return self
+end
+
+function StoryPreviewer:setDisplayParent(newParent: Instance)
+	self.display.Parent = newParent
+end
+
+function StoryPreviewer:setError(err)
+	local id = self.errorId + 1
+	self.errorId = id
+	task.delay(ERROR_DELAY, function()
+		if self.errorId ~= id then
+			-- Error was canceled or replaced.
+			return
+		end
+		warn(err)
+	end)
+end
+
+function StoryPreviewer:cancelError()
+	self.errorId += 1
+end
+
+function StoryPreviewer:setExpand(expand: boolean)
+	self.expand = expand
+end
+
+function StoryPreviewer:setSelectedStory(selectedStory)
+	self.selectedStory = selectedStory
+end
+
+function StoryPreviewer:updateDisplay()
+	if not self.currentPreview then
+		return
+	end
+	local target = self.currentPreview.target
+	if not target then
+		return
+	end
+	if self.expand then
+		target.Parent = self.display
+	else
+		target.Parent = self.rootRef.current
+	end
+end
+
+function StoryPreviewer:refreshPreview()
+	local selectedStory = self.selectedStory
+	if not selectedStory then
+		self:clearPreview()
+		return
+	end
+	local err, nextState = self:prepareState(selectedStory)
+	if err then
+		self:setError(err)
+		return
+	end
+	self:clearPreview()
+	self.currentPreview = nextState
+	self.setCurrentPreview(self.currentPreview)
+	self:updateDisplay()
+end
+
+function StoryPreviewer:clearPreview()
+	self:cancelError()
+	local state = self.currentPreview
+	if state == nil then
+		return
+	end
+	state:destroy()
+	self.currentPreview = nil
+	self.setCurrentPreview(self.currentPreview)
+end
+
+function StoryPreviewer:prepareState(selectedStory)
+	local state = {
+		cleanup = nil,
+		monkeyRequireCache = {},
+		monkeyGlobalTable = {},
+		scriptChangedConnections = {},
+		target = nil,
+	}
+
+	function state:destroy()
```

Comment:
Selene will warn about shadowing self. Functions here don't require `:` since you already have direct access to the state table.